### PR TITLE
Garbage-collect old receipts per block

### PR DESCRIPTION
**Pull request by liayoo**
Updated to garage-collect receipts per block instead of per transaction for performance.

### CODE REVIEW

File: consensus/index.js [hunk 1]
```diff
@@ -703,6 +704,7 @@ class Consensus {
         level: 'error'
       });
     }
+    newDb.removeOldReceipts();
```
Comment (platfowner) [inline consensus/index.js:707]: It's not very intuitive to me to call removeOldReceipts() in validateTransaction(). 
Is it possible to have a centralized module for executing block on db (e.g. executeBlockOnDb()) and use it everywhere?

File: consensus/index.js [hunk 2]
```diff
@@ -421,6 +421,7 @@ class Consensus {
         validators, recordedInvalidBlockHashSet, blockTime, tempDb);
     const { transactions, receipts, gasAmountTotal, gasCostTotal } = this.getValidTransactions(
         longestNotarizedChain, blockNumber, blockTime, tempDb);
+    tempDb.removeOldReceipts();
```
Comment (platfowner) [inline consensus/index.js:424]: ditto

File: consensus/index.js (hunk 1 above)
Comment (liayoo) [reply consensus/index.js]: Right, thanks for the idea! I've added a function called validateAndExecuteBlockOnDb() in Consensus that performs block validation & execution. PTAL~

File: consensus/index.js [hunk 3]
```diff
@@ -656,29 +662,31 @@ class Consensus {
...
+            level: 'error'
+          });
+        }
       }
     }
   }
 
-  validateTransactions(transactions, receipts, number, blockTime, expectedGasAmountTotal, expectedGasCostTotal, newDb) {
-    const txsRes = newDb.executeTransactionList(transactions, false, true, number, blockTime);
+  static validateTransactions(transactions, receipts, number, blockTime, expectedGasAmountTotal, expectedGasCostTotal, db) {
```
Comment (platfowner) [inline consensus/index.js:678]: what about validateAndExecuteTransactions() as it's actually executing the transactions on the given db?

File: tools/proof-hash/verifyBlock.js [hunk 4]
```diff
@@ -37,24 +39,11 @@ async function verifyBlock(snapshotFile, blockFileList) {
...
-            logger.error(`  Failed to execute evidence (${block.number})`);
-            process.exit(0);
-          }
-        }
-      }
-    }
-    if (!db.executeTransactionList(block.transactions, block.number === 0, true, block.number, block.timestamp)) {
-      logger.error(`  Failed to execute transactions (${block.number})`)
+    try {
+      Consensus.validateAndExecuteBlockOnDb(block, snapDb, prevBlock);
```
Comment (platfowner) [inline tools/proof-hash/verifyBlock.js:43]: nice modularization!

File: node/index.js [hunk 5]
```diff
@@ -592,51 +570,17 @@ class BlockchainNode {
...
-
-      // NOTE(platfowner): Write the current snapshot for debugging.
-      const snapshot = db.stateRoot.toJsObject();
-      FileUtil.writeSnapshot(this.snapshotDir, block.number, snapshot, true);
-
-      // NOTE(liayoo): Quick fix for the problem. May be fixed by deleting the block files.
+  executeBlockOnDbAndCleanUp(block, db, prevBlock) {
+    const LOG_HEADER = 'executeBlockOnDbAndCleanUp';
+    try {
+      Consensus.validateAndExecuteBlockOnDb(block, db, prevBlock);
```
Comment (platfowner) [inline node/index.js:576]: nice!

File: consensus/index.js (hunk 2 above)
Comment (platfowner) [reply consensus/index.js:424]: What about moving this one level under, i.e. in this.getValidTransactions()?

Comment (liayoo) [reply consensus/index.js]: Oh, that's really nice! Thanks! Done.

File: consensus/index.js (hunk 3 above)
Comment (liayoo) [reply consensus/index.js]: Sounds good to me! Updated.

File: tools/proof-hash/verifyBlock.js (hunk 4 above)
Comment (liayoo) [reply tools/proof-hash/verifyBlock.js:43]: Thank you! :)

**Comment by liayoo** [comment]
Addressed all comments, PTAL @platfowner

**Review by platfowner — APPROVED** [review]
LGTM
Thanks for the nice optimization and refactoring.

**Pull request merged**